zboxcore/sdk/playlist.go

```diff
@@ -0,0 +1,190 @@
+package sdk
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/0chain/gosdk/core/encryption"
+	"github.com/0chain/gosdk/core/resty"
+	"github.com/0chain/gosdk/core/sys"
+	"github.com/0chain/gosdk/zboxcore/client"
+	"github.com/0chain/gosdk/zboxcore/logger"
+	"github.com/0chain/gosdk/zboxcore/zboxutil"
+)
+
+type PlaylistFile struct {
+	LookupHash string `gorm:"column:lookup_hash" json:"lookup_hash"`
+	Name       string `gorm:"column:name" json:"name"`
+	Path       string `gorm:"column:path" json:"path"`
+	NumBlocks  int64  `gorm:"column:num_of_blocks" json:"num_of_blocks"`
+	ParentPath string `gorm:"column:parent_path" json:"parent_path"`
+	Size       int64  `gorm:"column:size;" json:"size"`
+	MimeType   string `gorm:"column:mimetype" json:"mimetype"`
+	Type       string `gorm:"column:type" json:"type"`
+}
+
+func GetPlaylist(ctx context.Context, alloc *Allocation, path, since string) ([]PlaylistFile, error) {
+
+	q := &url.Values{}
+	q.Add("path", path)
+	q.Add("since", since)
+
+	return getPlaylistFromBlobbers(ctx, alloc, q.Encode())
+}
+
+func GetPlaylistByAuthTicket(ctx context.Context, alloc *Allocation, authTicket, lookupHash, since string) ([]PlaylistFile, error) {
+
+	q := &url.Values{}
+	q.Add("auth_token", authTicket)
+	q.Add("lookup_hash", lookupHash)
+	q.Add("since", since)
+
+	return getPlaylistFromBlobbers(ctx, alloc, q.Encode())
+}
+
+func getPlaylistFromBlobbers(ctx context.Context, alloc *Allocation, query string) ([]PlaylistFile, error) {
+
+	urls := make([]string, len(alloc.Blobbers))
+	for i, b := range alloc.Blobbers {
+		sb := &strings.Builder{}
+		sb.WriteString(strings.TrimRight(b.Baseurl, "/"))
+		sb.WriteString(zboxutil.PLAYLIST_ENDPOINT)
+		sb.WriteString(alloc.ID)
+		sb.WriteString("?")
+		sb.WriteString(query)
+
+		urls[i] = sb.String()
+	}
+
+	opts := make([]resty.Option, 0, 3)
+
+	opts = append(opts, resty.WithRetry(resty.DefaultRetry))
+	opts = append(opts, resty.WithTimeout(resty.DefaultRequestTimeout))
+	opts = append(opts, resty.WithRequestInterceptor(func(req *http.Request) {
+		req.Header.Set("X-App-Client-ID", client.GetClientID())
+		req.Header.Set("X-App-Client-Key", client.GetClientPublicKey())
+
+		hash := encryption.Hash(alloc.ID)
+		sign, err := sys.Sign(hash, client.GetClient().SignatureScheme, client.GetClientSysKeys())
+		if err != nil {
+			logger.Logger.Error("playlist: ", err)
```

Comment:
Should not we return error here?

Reply:
error is unsupported in current `WithRequestInterceptor`. do you think it should be changed?

```
// WithRequestInterceptor intercept request
func WithRequestInterceptor(interceptor func(req *http.Request)) Option {
	return func(r *Resty) {
		r.requestInterceptor = interceptor
	}
}
```

Reply:
hmm, if the sign result is a must, then we should check and return if error happens

Reply:
@peterlimg  fixed

Reply:
Seems the request will be sent out anyway, though the response will be ignored. So I would suggest to move the siging code out of the option function so that we can check and return early, we should only need to set the signature in the option function.

Reply:
fixed